ports/nrf/common-hal/pulseio/PulseIn.c

```diff
@@ -35,50 +36,247 @@
+    if (self->buffer == NULL) {
+        mp_raise_msg_varg(&mp_type_MemoryError, translate("Failed to allocate RX buffer of %d bytes"), maxlen * sizeof(uint16_t));
+    }
+
+    self->pin = pin->number;
+    self->maxlen = maxlen;
+    self->idle_state = idle_state;
+    self->start = 0;
+    self->len = 0;
+    self->first_edge = true;
+    self->paused = false;
+    self->last_us = 0;
+    self->last_ms = 0;
+
+    claim_pin(pin);
+
+    nrfx_gpiote_in_config_t cfg = {
+        .sense = NRF_GPIOTE_POLARITY_TOGGLE,
+        .pull = NRF_GPIO_PIN_NOPULL, // idle_state ? NRF_GPIO_PIN_PULLDOWN : NRF_GPIO_PIN_PULLUP,
+        .is_watcher = false, // nrf_gpio_cfg_watcher vs nrf_gpio_cfg_input
+        .hi_accuracy = true,
+        .skip_gpio_setup = false
+    };
+    nrfx_gpiote_in_init(self->pin, &cfg, _pulsein_handler);
+    nrfx_gpiote_in_event_enable(self->pin, true);
 }
 
 bool common_hal_pulseio_pulsein_deinited(pulseio_pulsein_obj_t* self) {
-  return 1;
+  return self->pin == NO_PIN;
 }
 
 void common_hal_pulseio_pulsein_deinit(pulseio_pulsein_obj_t* self) {
+    if (common_hal_pulseio_pulsein_deinited(self)) {
+        return;
+    }
+
+    nrfx_gpiote_in_event_disable(self->pin);
+    nrfx_gpiote_in_uninit(self->pin);
+
+    // mark local array as invalid
+    int idx = _find_pulsein_obj(self);
+    if ( idx < 0 ) {
+        mp_raise_NotImplementedError(NULL);
+    }
+    _objs[idx] = NULL;
 
+    reset_pin_number(self->pin);
+    self->pin = NO_PIN;
 }
 
 void common_hal_pulseio_pulsein_pause(pulseio_pulsein_obj_t* self) {
-
+    nrfx_gpiote_in_event_disable(self->pin);
+    self->paused = true;
 }
 
 void common_hal_pulseio_pulsein_resume(pulseio_pulsein_obj_t* self, uint16_t trigger_duration) {
+    // Make sure we're paused.
+    if ( !self->paused ) {
+        common_hal_pulseio_pulsein_pause(self);
+    }
+
+    // Send the trigger pulse.
+    if (trigger_duration > 0) {
+        nrfx_gpiote_in_uninit(self->pin);
+
+        nrf_gpio_cfg_output(self->pin);
+        nrf_gpio_pin_write(self->pin, !self->idle_state);
+        common_hal_mcu_delay_us((uint32_t)trigger_duration);
+        nrf_gpio_pin_write(self->pin, self->idle_state);
+
+        nrfx_gpiote_in_config_t cfg = {
+            .sense = NRF_GPIOTE_POLARITY_TOGGLE,
+            .pull = NRF_GPIO_PIN_NOPULL, // idle_state ? NRF_GPIO_PIN_PULLDOWN : NRF_GPIO_PIN_PULLUP,
+            .is_watcher = false, // nrf_gpio_cfg_watcher vs nrf_gpio_cfg_input
+            .hi_accuracy = true,
+            .skip_gpio_setup = false
+        };
+        nrfx_gpiote_in_init(self->pin, &cfg, _pulsein_handler);
+    }
+
+    self->first_edge = true;
+    self->paused = false;
+    self->last_ms = 0;
+    self->last_us = 0;
 
+    nrfx_gpiote_in_event_enable(self->pin, true);
 }
 
 void common_hal_pulseio_pulsein_clear(pulseio_pulsein_obj_t* self) {
+    if ( !self->paused ) {
+        nrfx_gpiote_in_event_disable(self->pin);
+    }
```

Comment:
Since this directly disables pulsein, it does not set self.paused == true  - Is that intentional?